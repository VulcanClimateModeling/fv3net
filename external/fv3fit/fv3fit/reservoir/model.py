import fsspec
import numpy as np
import os
from typing import Iterable, Hashable, Sequence, cast
import xarray as xr
import yaml

import fv3fit
from fv3fit import Predictor
from .readout import ReservoirComputingReadout
from .reservoir import Reservoir
from .domain2 import RankXYDivider
from fv3fit._shared import io
from .utils import square_even_terms
from .transformers import ReloadableTransfomer, encode_columns, decode_columns


def _transpose_xy_dims(ds: xr.Dataset, rank_dims: Sequence[str]):
    # Useful for transposing the x, y dims in a dataset to match those in
    # RankDivider.rank_dims, and leaves other dims in the same order
    # relative to x,y. Dims after the first occurence of one of the rank_dims
    # are assumed to be feature dims.
    # e.g. (time, y, x, z) -> (time, x, y, z) for rank_dims=(x, y)
    leading_non_xy_dims = []
    for dim in ds.dims:
        if dim not in rank_dims:
            leading_non_xy_dims.append(dim)
        if dim in rank_dims:
            break
    ordered_dims = (*leading_non_xy_dims, *rank_dims)
    return ds.transpose(*ordered_dims, ...)


@io.register("hybrid-reservoir")
class HybridReservoirComputingModel(Predictor):
    _HYBRID_VARIABLES_NAME = "hybrid_variables.yaml"

    def __init__(
        self,
        input_variables: Iterable[Hashable],
        hybrid_variables: Iterable[Hashable],
        output_variables: Iterable[Hashable],
        reservoir: Reservoir,
        readout: ReservoirComputingReadout,
        rank_divider: RankXYDivider,
        autoencoder: ReloadableTransfomer,
        square_half_hidden_state: bool = False,
    ):
        # TODO: The autoencoder and by  extension the rank encoder all assume
        # that the same variable set are used for inputs, hybrid variables, and
        # perhaps outputs.  This will quickly not be the case, so need to allow
        # for different ones for each case.  Default can be an expectation that
        # all variables are available, but the encoder will then contain all that
        # information (e.g., time=t and t+1).  Separate PR is necessary for that.
        self.reservoir_model = ReservoirComputingModel(
            input_variables=input_variables,
            output_variables=output_variables,
            reservoir=reservoir,
            readout=readout,
            square_half_hidden_state=square_half_hidden_state,
            rank_divider=rank_divider,
            autoencoder=autoencoder,
        )
        self.input_variables = input_variables
        self.hybrid_variables = hybrid_variables
        self.output_variables = output_variables
        self.readout = readout
        self.square_half_hidden_state = square_half_hidden_state
        self.rank_divider = rank_divider
        self.autoencoder = autoencoder

        self._no_overlap_divider = self.rank_divider.get_no_overlap_rank_divider()

    def predict(self, hybrid_input: Sequence[np.ndarray]):
        # hybrid input is assumed to be in original spatial xy dims
        # (x, y, feature) and does not include overlaps.
        encoded_hybrid_input = encode_columns(
            input_arrs=hybrid_input, transformer=self.autoencoder
        )

        flat_hybrid_in = self._no_overlap_divider.get_all_subdomains_with_flat_feature(
            encoded_hybrid_input
        )
        readout_input = self._concatenate_readout_inputs(
            self.reservoir_model.reservoir.state, flat_hybrid_in
        )

        flat_prediction = self.readout.predict(readout_input)
        prediction = self._no_overlap_divider.merge_all_flat_feature_subdomains(
            flat_prediction
        )
        decoded_prediction = decode_columns(
            encoded_output=prediction, transformer=self.autoencoder,
        )
        return decoded_prediction

    def _concatenate_readout_inputs(self, hidden_state_input, flat_hybrid_input):
        # TODO: square even terms should be handled by Reservoir model?
        if self.square_half_hidden_state is True:
            hidden_state_input = square_even_terms(hidden_state_input, axis=-1)

        readout_input = np.concatenate([hidden_state_input, flat_hybrid_input], axis=-1)
        return readout_input

    def reset_state(self):
        self.reservoir_model.reset_state()

    def increment_state(self, prediction_with_overlap: Sequence[np.ndarray]) -> None:
        self.reservoir_model.increment_state(prediction_with_overlap)

    def synchronize(self, synchronization_time_series):
        self.reservoir_model.synchronize(synchronization_time_series)

    def dump(self, path: str) -> None:
        self.reservoir_model.dump(path)
        with fsspec.open(os.path.join(path, self._HYBRID_VARIABLES_NAME), "w") as f:
            f.write(yaml.dump({"hybrid_variables": self.hybrid_variables}))

    @classmethod
    def load(cls, path: str) -> "HybridReservoirComputingModel":
        pure_reservoir_model = ReservoirComputingModel.load(path)
        with fsspec.open(os.path.join(path, cls._HYBRID_VARIABLES_NAME), "r") as f:
            hybrid_variables = yaml.safe_load(f)["hybrid_variables"]
        return cls(
            input_variables=pure_reservoir_model.input_variables,
            output_variables=pure_reservoir_model.output_variables,
            reservoir=pure_reservoir_model.reservoir,
            readout=pure_reservoir_model.readout,
            square_half_hidden_state=pure_reservoir_model.square_half_hidden_state,
            rank_divider=pure_reservoir_model.rank_divider,
            autoencoder=pure_reservoir_model.autoencoder,
            hybrid_variables=hybrid_variables,
        )


class HybridDatasetAdapter:
    def __init__(self, model: HybridReservoirComputingModel) -> None:
        self.model = model

    def predict(self, inputs: xr.Dataset) -> xr.Dataset:
        # TODO: potentially use in train.py instead of special functions there
        xy_input_arrs = self._input_dataset_to_arrays(inputs)  # x, y, feature dims

        prediction_arr = self.model.predict(xy_input_arrs)
        return self._output_array_to_ds(prediction_arr, dims=list(inputs.dims))

    def increment_state(self, inputs: xr.Dataset):
        xy_input_arrs = self._input_dataset_to_arrays(inputs)  # x, y, feature dims
        self.model.increment_state(xy_input_arrs)

    def reset_state(self):
        self.model.reset_state()

    def _input_dataset_to_arrays(self, inputs: xr.Dataset) -> Sequence[np.ndarray]:
        # Converts from xr dataset to sequence of variable ndarrays expected by encoder
        # Make sure the xy dimensions match the rank divider
        transposed_inputs = _transpose_xy_dims(
            ds=inputs, rank_dims=self.model.rank_divider._rank_dims_all_features
        )
        input_arrs = []
        for variable in self.model.input_variables:
            da = transposed_inputs[variable]
            if "z" not in da.dims:
                da = da.expand_dims("z", axis=-1)
            input_arrs.append(da.values)
        return input_arrs

    def _output_array_to_ds(
        self, outputs: Sequence[np.ndarray], dims: Sequence[str]
    ) -> xr.Dataset:
        ds = xr.Dataset(
            {
                var: (dims, outputs[i])
                for i, var in enumerate(self.model.output_variables)
            }
        )
        return ds


@io.register("pure-reservoir")
class ReservoirComputingModel(Predictor):
    _RESERVOIR_SUBDIR = "reservoir"
    _READOUT_SUBDIR = "readout"
    _METADATA_NAME = "metadata.yaml"
    _RANK_DIVIDER_NAME = "rank_divider.yaml"
    _AUTOENCODER_SUBDIR = "autoencoder"

    def __init__(
        self,
        input_variables: Iterable[Hashable],
        output_variables: Iterable[Hashable],
        reservoir: Reservoir,
        readout: ReservoirComputingReadout,
        rank_divider: RankXYDivider,
        autoencoder: ReloadableTransfomer,
        square_half_hidden_state: bool = False,
    ):
        """_summary_

        Args:
            reservoir: Reservoir which takes input and updates hidden state
            readout: readout layer which takes in state and predicts next time step
            square_half_hidden_state: if True, square even terms in the reservoir
                state before it is used as input to the regressor's .fit and
                .predict methods. This option was found to be important for skillful
                predictions in Wikner+2020 (https://doi.org/10.1063/5.0005541).
            rank_divider: object used to divide and reconstruct domain <-> subdomains
        """
        self.input_variables = input_variables
        self.output_variables = output_variables
        self.reservoir = reservoir
        self.readout = readout
        self.square_half_hidden_state = square_half_hidden_state
        self.rank_divider = rank_divider
        self.autoencoder = autoencoder

        self._no_overlap_divider = rank_divider.get_no_overlap_rank_divider()

    def process_state_to_readout_input(self):
        readout_input = self.reservoir.state
        if self.square_half_hidden_state is True:
            readout_input = square_even_terms(readout_input, axis=0)

        return readout_input

    def predict(self):
        # Returns raw readout prediction of latent state.
        readout_input = self.process_state_to_readout_input()
        flat_prediction = self.readout.predict(readout_input)
        prediction = self._no_overlap_divider.merge_all_flat_feature_subdomains(
            flat_prediction
        )
        decoded_prediction = decode_columns(
            encoded_output=prediction, transformer=self.autoencoder,
        )
        return decoded_prediction

    def reset_state(self):
        input_shape = (
            self.rank_divider.n_subdomains,
            self.reservoir.hyperparameters.state_size,
        )
        self.reservoir.reset_state(input_shape)

    def increment_state(self, prediction_with_overlap: Sequence[np.ndarray]) -> None:
        # input array is in native x, y, z_feature coordinates
        encoded_xy_input_arrs = encode_columns(
            prediction_with_overlap, self.autoencoder
        )
        encoded_flat_sub = self.rank_divider.get_all_subdomains_with_flat_feature(
            encoded_xy_input_arrs
        )
        self.reservoir.increment_state(encoded_flat_sub)

    def synchronize(self, synchronization_time_series):
        # input arrays in native x, y, z_feature coordinates
        encoded_timeseries = encode_columns(
            synchronization_time_series, self.autoencoder
        )
        encoded_flat = self.rank_divider.get_all_subdomains_with_flat_feature(
            encoded_timeseries
        )
        self.reservoir.synchronize(encoded_flat)

    def dump(self, path: str) -> None:
        """Dump data to a directory

        Args:
            path: a URL pointing to a directory
        """
        self.reservoir.dump(os.path.join(path, self._RESERVOIR_SUBDIR))
        self.readout.dump(os.path.join(path, self._READOUT_SUBDIR))

        metadata = {
            "square_half_hidden_state": self.square_half_hidden_state,
            "input_variables": self.input_variables,
            "output_variables": self.output_variables,
        }
        with fsspec.open(os.path.join(path, self._METADATA_NAME), "w") as f:
            f.write(yaml.dump(metadata))

        self.rank_divider.dump(os.path.join(path, self._RANK_DIVIDER_NAME))
        if self.autoencoder is not None:
            fv3fit.dump(self.autoencoder, os.path.join(path, self._AUTOENCODER_SUBDIR))

    @classmethod
    def load(cls, path: str) -> "ReservoirComputingModel":
        """Load a model from a remote path"""
        reservoir = Reservoir.load(os.path.join(path, cls._RESERVOIR_SUBDIR))
        readout = ReservoirComputingReadout.load(
            os.path.join(path, cls._READOUT_SUBDIR)
        )
        with fsspec.open(os.path.join(path, cls._METADATA_NAME), "r") as f:
            metadata = yaml.safe_load(f)

        rank_divider = RankXYDivider.load(os.path.join(path, cls._RANK_DIVIDER_NAME))

        autoencoder = cast(
            ReloadableTransfomer,
            fv3fit.load(os.path.join(path, cls._AUTOENCODER_SUBDIR)),
        )
        return cls(
            input_variables=metadata["input_variables"],
            output_variables=metadata["output_variables"],
            reservoir=reservoir,
            readout=readout,
            square_half_hidden_state=metadata["square_half_hidden_state"],
            rank_divider=rank_divider,
            autoencoder=autoencoder,
        )
