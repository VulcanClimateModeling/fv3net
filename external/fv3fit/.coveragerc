[run]
omit = tests/* setup.py fv3fit/sklearn/__main__.py fv3fit/keras/__main__.py
