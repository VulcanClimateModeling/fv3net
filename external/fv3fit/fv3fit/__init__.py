from ._shared.predictor import Predictor, Estimator
from ._shared.io import dump, load
from . import keras
from . import sklearn