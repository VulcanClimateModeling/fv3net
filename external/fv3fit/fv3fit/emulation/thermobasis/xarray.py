import dataclasses
import logging
from typing import MutableMapping, Hashable, Union
import xarray as xr
import tensorflow as tf
from fv3fit.emulation.thermobasis.batch import to_tensors, to_dict_no_static_vars
from fv3fit.emulation.thermobasis.emulator import Trainer, Config

State = MutableMapping[Hashable, xr.DataArray]


@dataclasses.dataclass
class XarrayEmulator:
    """Wrap an OnlineEmulator to allow for xarrray inputs and outputs
    """

    emulator: Trainer

    @property
    def input_variables(self):
        return self.emulator.input_variables

    @property
    def output_variables(self):
        return self.emulator.output_variables

    def dump(self, path: str):
        return self.emulator.dump(path)

    @classmethod
    def load(cls, path: str):
        return cls(Trainer.load(path))

    def predict(self, state: State) -> State:
        in_ = stack(state, self.input_variables)
        in_tensors = to_tensors(in_)
        x = self.emulator.batch_to_specific_humidity_basis(in_tensors)
        out = self.emulator.model(x)

        tensors = to_dict_no_static_vars(out)

        dims = ["sample", "z"]
        attrs = {"units": "no one cares"}

        return dict(
            xr.Dataset(
                {key: (dims, val, attrs) for key, val in tensors.items()},
                coords=in_.coords,
            ).unstack("sample")
        )

    def partial_fit(self, statein: State, stateout: State):

        in_tensors = _xarray_to_tensor(statein, self.input_variables)
        out_tensors = _xarray_to_tensor(stateout, self.output_variables)
        d = tf.data.Dataset.from_tensor_slices((in_tensors, out_tensors)).shuffle(
            1_000_000
        )
        self.emulator.batch_fit(d)


def _xarray_to_tensor(state, keys):
    in_ = stack(state, keys)
    return to_tensors(in_)


def stack(state: State, keys) -> xr.Dataset:
    ds = xr.Dataset({key: state[key] for key in keys})
    sample_dims = ["y", "x"]
    return ds.stack(sample=sample_dims).transpose("sample", ...)


def get_xarray_emulator(config: Union[Config, str]) -> XarrayEmulator:
    if isinstance(config, str):
        logging.info(f"Loading emulator from checkpoint {config}")
        return XarrayEmulator(Trainer.load(config))
    else:
        return XarrayEmulator(Trainer(config))
