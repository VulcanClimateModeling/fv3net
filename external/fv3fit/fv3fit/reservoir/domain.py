import numpy as np


def slice_along_axis(arr: np.ndarray, inds: slice, axis: int = 0):
    # https://stackoverflow.com/a/37729566
    # For slicing ndarray along a dynamically specified axis
    # same as np.take() but does not make a copy of the data
    sl = [slice(None)] * arr.ndim
    sl[axis] = inds
    return arr[tuple(sl)]
