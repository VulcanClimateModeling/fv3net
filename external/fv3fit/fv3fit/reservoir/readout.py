import dacite
import dataclasses
import fsspec
import io
import joblib
import numpy as np
import re
import scipy.sparse
from typing import Sequence, Optional

from .config import ReadoutHyperparameters, BatchLinearRegressorHyperparameters


class NotFittedError(Exception):
    def __init__(self, message):
        super().__init__(message)


def _square_evens(v: np.ndarray) -> np.ndarray:
    evens = v[::2]
    odds = v[1::2]
    c = np.empty((v.size,), dtype=v.dtype)
    c[0::2] = evens ** 2
    c[1::2] = odds
    return c


def square_even_terms(v: np.ndarray, axis=1) -> np.ndarray:
    return np.apply_along_axis(func1d=_square_evens, axis=axis, arr=v)


def _extract_int_from_subdir(s):
    nums = re.findall(r"\d+", s.rstrip("/").split("/")[-1])
    if len(nums) == 1:
        return int(nums[0])
    else:
        raise ValueError(
            "Ordering of readout subdirectories should be indicated "
            "by a single numeric tag or suffix. ex. 'subdir_0' or '0'."
            f"Subdir {s} violates this naming rule."
        )


def _sort_subdirs_numerically(subdirs: Sequence[str]) -> Sequence[str]:
    """
    Sort a list of subdirs by their numeric tags.
    ex. "subdir_0", "subdir_1"
    """
    nums = [_extract_int_from_subdir(s) for s in subdirs]
    if len(subdirs) != len(np.unique(nums)):
        raise ValueError(
            "Multiple readout subdirectories have the same " "numbering label."
        )
    return [subdir for _, subdir in sorted(zip(nums, subdirs))]


class BatchLinearRegressor:
    """ Solves for weights W in matrix equation AW=B
    Where A = X_T.X + l2*I and B= X_T.y
    """

    def __init__(self, hyperparameters: BatchLinearRegressorHyperparameters):
        self.hyperparameters = hyperparameters
        self.A = None
        self.B = None

    def _add_bias_feature(self, X):
        return np.concatenate([X, np.ones((X.shape[0], 1))], axis=1)

    def _check_X_last_col_constant(self, X):
        last_col = X[:, -1]
        if not np.allclose(np.unique(last_col), np.array([1.0])):
            raise ValueError(
                "Last column of X array must all be ones if add_bias_term " "is False."
            )

    def batch_update(self, X, y):
        if self.hyperparameters.add_bias_term:
            X = self._add_bias_feature(X)
        else:
            self._check_X_last_col_constant(X)

        if self.A is None and self.B is None:
            self.A = np.zeros((X.shape[1], X.shape[1]))
            self.B = np.zeros((X.shape[1], y.shape[1]))

        self.A = np.add(self.A, (np.dot(X.T, X)))
        self.B = np.add(self.B, np.dot(X.T, y))

    def get_weights(self):
        # use_least_squares_solve is useful for simple test cases
        # where np.linalg.solve encounters for singular XT.X

        reg = self.hyperparameters.l2 * np.identity(self.A.shape[1])

        if self.A is None and self.B is None:
            raise NotFittedError(
                "At least one call of batch_update on data must be done "
                "before solving for weights."
            )
        if self.hyperparameters.use_least_squares_solve:
            W = np.linalg.lstsq(self.A + reg, self.B)[0]
        else:
            W = np.linalg.solve(self.A + reg, self.B)
        coefficients, intercepts = W[:-1, :], W[-1, :]
        return coefficients, intercepts


class ReservoirComputingReadout:
    """Readout layer of the reservoir computing model

    hyperparameters: hyperparameters describing the readout
    coefficients: if provided from an already-fit readout,
        use as the linear regression coefficients
    intercepts: if provided from an already-fit readout,
        use as the linear regression intercepts
    """

    _READOUT_NAME = "readout.bin"
    _COEFFICIENTS_NAME = "readout_coefficients.npz"

    def __init__(
        self,
        hyperparameters: ReadoutHyperparameters,
        coefficients: Optional[np.ndarray] = None,
        intercepts: Optional[np.ndarray] = None,
    ):
        self.hyperparameters = hyperparameters
        self.coefficients = coefficients
        self.intercepts = intercepts
        self.square_half_hidden_state = hyperparameters.square_half_hidden_state
        self.linear_regressor = BatchLinearRegressor(
            hyperparameters.linear_regressor_config
        )

    # TODO: Remove this method from readout and directly call square_even_terms and
    # BatchLinearRegressor.batch_update in the training function. The saved
    # Predictor and its components should not have a fit method.
    def fit(
        self,
        res_states: np.ndarray,
        output_states: np.ndarray,
        calculate_weights: bool = True,
    ) -> None:
        if self.square_half_hidden_state is True:
            res_states = square_even_terms(res_states, axis=1)
        self.linear_regressor.batch_update(res_states, output_states)
        if calculate_weights:
            self.calculate_weights()

    def calculate_weights(self) -> None:
        self.coefficients, self.intercepts = self.linear_regressor.get_weights()

    def predict(self, input: np.ndarray):
        if self.square_half_hidden_state:
            input = square_even_terms(input, axis=0)
        if len(input.shape) > 1:
            print(f"input shape before flattening: {input.shape}")
            input = input.reshape(-1)
            print(f"input shape after flattening: {input.shape}")

        flat_prediction = (input * self.coefficients) + self.intercepts
        print(f"flat_prediction shape: {flat_prediction.shape}")

        if len(input.shape) > 1:
            return flat_prediction.reshape(-1, input.shape[-1])
        else:
            return flat_prediction

    def dump(self, path: str) -> None:
        """Dump data to a directory

        Args:
            path: a URL pointing to a directory
        """

        fs: fsspec.AbstractFileSystem = fsspec.get_fs_token_paths(path)[0]
        fs.makedirs(path, exist_ok=True)
        mapper = fs.get_mapper(path)
        components = {
            "lr_hyperparameters": dataclasses.asdict(
                self.hyperparameters.linear_regressor_config
            ),
            "square_half_hidden_state": self.hyperparameters.square_half_hidden_state,
            # "coefficients": self.coefficients,
            "intercepts": self.intercepts,
        }
        f = io.BytesIO()
        joblib.dump(components, f)
        mapper[self._READOUT_NAME] = f.getvalue()
        with fs.open(f"{path}/{self._COEFFICIENTS_NAME}", "wb") as f:
            scipy.sparse.save_npz(f, self.coefficients)

    @classmethod
    def load(cls, path: str) -> "ReservoirComputingReadout":
        mapper = fsspec.get_mapper(path)
        fs: fsspec.AbstractFileSystem = fsspec.get_fs_token_paths(path)[0]
        f = io.BytesIO(mapper[cls._READOUT_NAME])
        readout_components = joblib.load(f)
        lr_config = dacite.from_dict(
            BatchLinearRegressorHyperparameters,
            readout_components.pop("lr_hyperparameters"),
        )
        hyperparameters = ReadoutHyperparameters(
            linear_regressor_config=lr_config,
            square_half_hidden_state=readout_components.pop("square_half_hidden_state"),
        )
        with fs.open(f"{path}/{cls._COEFFICIENTS_NAME}", "rb") as f:
            coefficients = scipy.sparse.load_npz(f)
        return cls(
            hyperparameters=hyperparameters,
            coefficients=coefficients,
            **readout_components,
        )


def combine_readouts(readouts: Sequence[ReservoirComputingReadout]):
    intercepts, square_state_settings, lr_config_hashes = [], [], []
    for readout in readouts:
        # coefs.append(readout.coefficients)
        intercepts.append(readout.intercepts)
        square_state_settings.append(readout.square_half_hidden_state)
        lr_config_hashes.append(readout.hyperparameters.linear_regressor_config.hash)

    if len(np.unique(square_state_settings)) != 1:
        raise ValueError(
            "All readouts must have the same setting for square_half_hidden_state."
        )
    if len(np.unique(lr_config_hashes)) != 1:
        raise ValueError(
            "All readouts must have the same hyperparameters for BatchLinearRegressor."
        )

    # Merge the coefficient arrays of individual readouts into single
    # block diagonal matrix
    combined_coefficients = scipy.sparse.block_diag(
        [readout.coefficients for readout in readouts]
    )  # scipy.sparse.block_diag(coefs)

    # Concatenate the intercepts of individual readouts into single array
    combined_intercepts = np.concatenate(intercepts)

    hyperparameters = readouts[0].hyperparameters
    return ReservoirComputingReadout(
        hyperparameters=hyperparameters,
        coefficients=combined_coefficients,
        intercepts=combined_intercepts,
    )


class CombinedReservoirComputingReadout:
    """Combines readout layers of multiple reservoir computing models
    into a block diagonal readout layer, which can be used to predict over
    the combined domain of those models.
    Prediction is done on vector of concatenated reservoir states.

    linear_regressors: sequence of sklearn Ridge regressors, each corresponding
        to a subdomain
    square_half_hidden_state: if True, square even terms in the reservoir state
        before it is used as input to the regressor's .fit and .predict methods
        This option was found to be important for skillful predictions in
        Wikner+2020 (https://doi.org/10.1063/5.0005541)
    """

    _READOUT_NAME = "readout.bin"

    def __init__(self, readouts: Sequence[ReservoirComputingReadout]):
        self._combine_readouts(readouts)

    def _combine_readouts(self, readouts: Sequence[ReservoirComputingReadout]):
        coefs, intercepts, square_state_settings = [], [], []
        for readout in readouts:
            coefs.append(readout.coefficients)
            intercepts.append(readout.intercepts)
            square_state_settings.append(readout.square_half_hidden_state)

        # Merge the coefficient arrays of individual readouts into single
        # block diagonal matrix
        self.coefficients = scipy.sparse.block_diag(coefs)

        # Concatenate the intercepts of individual readouts into single array
        self.intercepts = np.concatenate(intercepts)

        if len(np.unique(square_state_settings)) != 1:
            raise ValueError(
                "All readouts must have the same setting for square_half_hidden_state."
            )
        self.square_half_hidden_state = square_state_settings[0]

    def predict(self, input: np.ndarray):
        if self.square_half_hidden_state:
            input = square_even_terms(input, axis=0)
        return input * self.coefficients + self.intercepts

    @classmethod
    def load(cls, path: str) -> "CombinedReservoirComputingReadout":
        """Load a model from a remote directory. Each subdir in path
        refers to the full reservoir model directory containing the saved
        readout to load within a subdir named "readout". ex.
        | path
        | -- model_0
        |    -- readout.bin
        |    -- reservoir
        | -- model_1
        |    -- readout.bin
              ...
        Assumes model subdirs are numbered in the order that they should
        be used in the combined readout.

        path: directory containing model subdirectories for each readout
        """
        fs: fsspec.AbstractFileSystem = fsspec.get_fs_token_paths(path)[0]
        subdirs = _sort_subdirs_numerically(fs.ls(path))
        if "gs" in fs.protocol:
            subdirs = [f"gs://{path}" for path in subdirs]

        readouts = []

        for subdir in subdirs:
            mapper = fsspec.get_mapper(subdir)

            f = io.BytesIO(mapper[cls._READOUT_NAME])
            readout_components = joblib.load(f)
            lr_config = dacite.from_dict(
                BatchLinearRegressorHyperparameters,
                readout_components.pop("lr_hyperparameters"),
            )
            hyperparameters = ReadoutHyperparameters(
                linear_regressor_config=lr_config,
                square_half_hidden_state=readout_components.pop(
                    "square_half_hidden_state"
                ),
            )
            readouts.append(
                ReservoirComputingReadout(
                    hyperparameters=hyperparameters, **readout_components
                )
            )

        return cls(readouts)
