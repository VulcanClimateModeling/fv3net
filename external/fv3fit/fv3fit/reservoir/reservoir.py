import dataclasses
import logging
import numpy as np
import scipy
from typing import Optional

logger = logging.getLogger(__name__)


def _random_uniform_sample_func(min, max):
    def _f(d):
        return np.random.uniform(min, max, size=d)

    return _f


def _random_uniform_sparse_matrix(m, n, sparsity, min=0, max=1):
    return scipy.sparse.random(
        m=m,
        n=n,
        density=1.0 - sparsity,
        data_rvs=_random_uniform_sample_func(min=0, max=1.0),
    )


@dataclasses.dataclass
class ReservoirHyperparameters:
    """Hyperparameters for reservoir

    input_size: Size of input vector
    state_size: Size of hidden state vector,
        W_res has shape state_size x state_size
    adjacency_matrix_sparsity: Fraction of elements in adjacency matrix
        W_res that are zero
    output_size: Optional: size of output vector. Can be smaller than input
        dimension if predicting on subdomains with overlapping
        input regions. Defaults to same as input_size
    spectral_radius: Largest absolute value eigenvalue of W_res.
        Larger values increase the memory of the reservoir.
    seed: Random seed for sampling
    input_coupling_sparsity: Fraction of elements in each row of W_in
        that are zero. Kept the same in all rows to ensure each input
        is equally connected into the reservoir. Defaults to 0.
    input_coupling_scaling: Scaling applied to W_in. Defaults to 1,
        where all elements are sampled from random uniform distribution
        [-1, 1]. Changing this affects relative weighting of reservoir memory
        versus the most recent state.
    """

    input_size: int
    state_size: int
    adjacency_matrix_sparsity: float
    spectral_radius: float
    output_size: Optional[int] = None
    seed: int = 0
    input_coupling_sparsity: float = 0.0
    input_coupling_scaling: float = 1.0

    def __post_init__(self):
        if not self.output_size:
            self.output_size = self.input_size


class Reservoir:
    def __init__(
        self, hyperparameters: ReservoirHyperparameters,
    ):
        self.hyperparameters = hyperparameters
        np.random.seed(self.hyperparameters.seed)
        self.W_in = self._generate_W_in()
        self.W_res = self._generate_W_res()
        self.state = np.zeros(self.hyperparameters.state_size)

    def increment_state(self, input):
        self.state = np.tanh(self.W_in * input + self.W_res * self.state)

    def reset_state(self):
        logger.info("Resetting reservoir state.")
        self.state = np.zeros(self.hyperparameters.state_size)

    def synchronize(self, synchronization_time_series):
        self.reset_state()
        for input in synchronization_time_series:
            self.increment_reservoir_state(input)

    def _generate_W_in(self):
        W_in_rows = []
        # Generate by row to ensure same number of connections per input element
        for k in range(self.hyperparameters.state_size):
            W_in_rows.append(
                _random_uniform_sparse_matrix(
                    m=1,
                    n=self.hyperparameters.input_size,
                    sparsity=self.hyperparameters.input_coupling_sparsity,
                    min=-self.hyperparameters.input_coupling_scaling,
                    max=self.hyperparameters.input_coupling_scaling,
                )
            )

        return scipy.sparse.vstack(W_in_rows)

    def _generate_W_res(self):
        W_res = _random_uniform_sparse_matrix(
            m=self.hyperparameters.state_size,
            n=self.hyperparameters.state_size,
            sparsity=self.hyperparameters.adjacency_matrix_sparsity,
            min=0,
            max=1,
        )
        largest_magnitude_eigval = scipy.sparse.linalg.eigs(
            W_res, return_eigenvectors=False, k=1, which="LM"
        ).item()
        scaling = self.hyperparameters.spectral_radius / abs(largest_magnitude_eigval)

        return scaling * W_res
