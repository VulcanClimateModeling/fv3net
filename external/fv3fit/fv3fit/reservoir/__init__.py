from .config import (
    ReservoirHyperparameters,
    CubedsphereSubdomainConfig,
    ReservoirTrainingConfig,
)
from .reservoir import Reservoir
from .readout import ReservoirComputingReadout, square_even_terms, combine_readouts
from .model import ReservoirComputingModel
from .domain import RankDivider, stack_time_series_samples
