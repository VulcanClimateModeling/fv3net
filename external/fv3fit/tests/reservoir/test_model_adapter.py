import numpy as np
import xarray as xr

from fv3fit.reservoir.transformers.transformer import Transformer
from fv3fit.reservoir.domain import RankDivider
from fv3fit.reservoir.readout import ReservoirComputingReadout
from fv3fit.reservoir import (
    Reservoir,
    ReservoirHyperparameters,
    HybridReservoirComputingModel,
)
from fv3fit.reservoir.model import HybridDatasetAdapter


class DoNothingAutoencoder(Transformer):
    def __init__(self, latent_dims):
        self._latent_dim_len = latent_dims

    @property
    def n_latent_dims(self):
        return self._latent_dim_len

    def encode(self, x):
        return np.concatenate(x, -1)

    def decode(self, latent_x):
        return latent_x


def get_initialized_hybrid_model():

    input_size = 2 * 2 * 6  # no overlap subdomain in latent space
    output_size = 2 * 2 * 6  # no overlap subdomain in latent space
    hybrid_input_size = 6 * 6 * 6  # overlap subdomain in latent space
    state_size = 25
    hyperparameters = ReservoirHyperparameters(
        state_size=state_size,
        adjacency_matrix_sparsity=0.0,
        spectral_radius=1.0,
        input_coupling_sparsity=1,
    )
    reservoir = Reservoir(hyperparameters, input_size=input_size)

    # multiplied by the number of subdomains since it's a combined readout
    readout = ReservoirComputingReadout(
        coefficients=np.random.rand(
            state_size * 4 + hybrid_input_size * 4, output_size * 4
        ),
        intercepts=np.random.rand(output_size * 4),
    )

    # expects rank size (including halos) in latent space
    divider = RankDivider((2, 2), ["x", "y", "z"], [8, 8, 6], 2)

    hybrid_predictor = HybridReservoirComputingModel(
        input_variables=["a", "b"],
        output_variables=["a", "b"],
        hybrid_variables=["a", "b"],
        reservoir=reservoir,
        readout=readout,
        rank_divider=divider,
        autoencoder=DoNothingAutoencoder(6),
    )
    hybrid_predictor.reset_state()

    return hybrid_predictor


def get_single_rank_xarray_data():
    rng = np.random.RandomState(0)
    a = rng.randn(8, 8, 3)  # two variables concatenated to form size 6 latent space
    b = rng.randn(8, 8, 3)

    return xr.Dataset(
        {
            "a": xr.DataArray(a, dims=["x", "y", "z"]),
            "b": xr.DataArray(b, dims=["x", "y", "z"]),
        }
    )


def test_adapter_predict(regtest):
    hybrid_predictor = get_initialized_hybrid_model()
    data = get_single_rank_xarray_data()

    model = HybridDatasetAdapter(hybrid_predictor)
    result = model.predict(data)

    print(result, file=regtest)
