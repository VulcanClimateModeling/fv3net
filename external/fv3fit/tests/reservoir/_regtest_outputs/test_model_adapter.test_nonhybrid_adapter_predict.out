<xarray.Dataset>
Dimensions:  (x: 4, y: 4, z: 3)
Dimensions without coordinates: x, y, z
Data variables:
    a        (x, y, z) float64 0.1522 -0.3741 -0.01345 ... 0.1408 0.4735 1.855
    b        (x, y, z) float64 0.8155 0.4106 0.481 ... 1.416 -0.3027 0.9897
